cmd/soroban-rpc/internal/db/ledgerentry.go

```diff
@@ -364,22 +365,24 @@ func (r ledgerEntryReader) NewCachedTx(ctx context.Context) (LedgerEntryReadTx,
 		stmtCache:              sq.NewStmtCache(txSession.GetTx()),
 		latestLedgerSeqCache:   r.db.cache.latestLedgerSeq,
 		ledgerEntryCacheReadTx: &cacheReadTx,
+		ledgerReader:           NewLedgerReader(r.db),
 		tx:                     txSession,
 		buffer:                 xdr.NewEncodingBuffer(),
 	}, nil
 }
 
 func (r ledgerEntryReader) NewTx(ctx context.Context) (LedgerEntryReadTx, error) {
 	txSession := r.db.Clone()
+	r.db.cache.RLock()
 	if err := txSession.BeginTx(ctx, &sql.TxOptions{ReadOnly: true}); err != nil {
 		return nil, err
 	}
-	r.db.cache.RLock()
 	defer r.db.cache.RUnlock()
```

Comment:
I think the `defer r.db.cache.RUnlock()` needs to be moved up as well. if there is an error in `txSession.BeginTx()` that will result in the the cache being permanently read locked

Reply:
also, now that `NewTx()` and `NewCachedTx()` are almost identical I think we can refactor the code so that they use a common helper function to reduce code duplication